### PR TITLE
page_cache: eliminate `PageWriteGuard`

### PR DESCRIPTION
**Pull request by problame**
Since my #4994 , the page cache doesn't support mutating pages anymore.

The remaining role of `PageWriteGuard` is that of a fill handler. We can use `FnOnce` closures (or in an aysnc  future: a `Future`) for that.

So, make the change now.

### CODE REVIEW

**Review by hlinnaka — COMMENTED** [review]
I don't know, the new interface doesn't feel very nice to me. I think the PageWriteGuard was a fine approach. Was the motivation just to simplify things? I like that this is less lines of code, but other than that I'm not sure it's really simpler.

If we turn this async in the future, we'll need to add an async variant that takes a Future, as you said. PageWriteGuard gives the caller more flexibility, and I don't think it's any more complicated in the callers either.

File: pageserver/src/page_cache.rs [hunk 1]
```diff
@@ -249,81 +248,16 @@ impl AsRef<[u8; PAGE_SZ]> for PageReadGuard<'_> {
...
+    MissFilled,
 }
 
 /// lock_for_write() return value
-pub enum WriteBufResult<'a> {
-    Found(PageWriteGuard<'a>),
-    NotFound(PageWriteGuard<'a>),
+pub enum WriteBufResult {
+    Already,
+    Inserted,
```
Comment (hlinnaka) [inline pageserver/src/page_cache.rs:260]: These need more comments and/or clearer names too, IMHO

**Comment by problame** [comment]
> I don't know, the new interface doesn't feel very nice to me. I think the PageWriteGuard was a fine approach. Was the motivation just to simplify things? I like that this is less lines of code, but other than that I'm not sure it's really simpler.

Motivation is the first point in https://github.com/neondatabase/neon/pull/5111

The second point in #5111 is addressed by my https://github.com/neondatabase/neon/pull/5122 which is stacked on top of this PR here.

**Comment by arpad-m** [comment]
FTR, as the `remove_mapping` function can stay sync, this PR isn't really required.  (originally I wanted to make `remove_mapping` async in #5023 but Christian asked me if it's really needed to be async).

Also, personally I prefer the approach to just set the counter to 0 to this PR, like suggested in https://github.com/neondatabase/neon/pull/5111.

**Comment by problame** [comment]
Also, for the eventual switch to io-uring, having an owned buffer handle is required. PageWriteGuard is fit for that purpose. So, let's leave it and close this PR.

**Pull request closed without merging**